doc/reference/trd104-syscalls.md

```diff
@@ -509,19 +509,19 @@ kernel returns `Failure` with an error code of `NODEVICE`.
 ---------------------------------
 
 The Read-Write Allow system call class is how a userspace process
-shares a buffer with the kernel that the kernel can read and write. 
-
-Calling a Read-Write Allow system call returns a buffer (address and
-length).  On the first call to a Read-Write Allow system call, the
-kernel returns a zero-length buffer. Subsequent successful calls to
-Read-Write Allow return the previous buffer passed.  The standard
-access model for allowed buffers is that userspace does not read or
-write a buffer that has been allowed: access to the memory is intended
-to be exclusive either to userspace or to the kernel. To regain access
-to a passed buffer, the process calls the same Read-Write Allow system
-call again, which, if successful, returns the buffer. It can do so
-with a zero-length buffer if it wishes to pass no memory to the
-kernel. 
+shares a buffer with the kernel that the kernel can read and write.
+
+The standard access model for allowed buffers is that userspace does
+not read or write a buffer that has been allowed: access to the memory
+is intended to be exclusive either to userspace or to the kernel. To
+regain access to a passed buffer, the process calls the same
+Read-Write Allow system call again, which, if successful, returns the
+buffer. It can do so with a zero-length buffer if it wishes to pass no
```

Comment:
I've deliberately structured the text to say that the kernel does not enforce the stated requirements way at the beginning. This was driven by the fact that I'm worried that introducing these buffer-swapping semantics here might lead people to think that this is indeed always a one-for-one swap in the Allow slot, which it isn't.

It's entirely fine to state this below, I agree it makes more sense, and it's the developer's fault for not reading all the way. However, it's tricky, given this section literally says "returns the buffer". Changing it to "returns _a_ buffer" better incorporates what ABI guarantees we give:

```suggestion
Read-Write Allow system call again, which, if successful, returns a
buffer. The process can do so with a zero-length buffer if it wishes to pass no
```

(Also resolves some possible ambiguity of who can pass a zero-length buffer)

Reply:
I think this isn't right. This text is intended to describe how userspace regains exclusive access, which is my calling `allow` again. By saying "returns a buffer", it suggests that it might not be the same buffer that was passed: "a" means "some  buffer" while "the" means "the same buffer."

Reply:
I agree with the "process" point. I did a bit of reworking, take a look.